File: go/342.go
/*
- Use logarithm, 4^x = num, x = log(num) / log(4)
- num & 0x55555555, 0b10101010... makes sure it is not power of 2
- 4^n - 1^n = (4-1)*(...) = 3*x
*/
func isPowerOfFour(num int) bool {
    if num <= 0 {
        return false
    }
    temp := math.Log(float64(num)) / math.Log(4)
    return temp == math.Trunc(temp)
}

func isPowerOfFour(num int) bool {
    return num > 0 && (num & (num-1)) == 0 && (num & 0x55555555) == num
}

func isPowerOfFour(num int) bool {
    return num > 0 && (num & (num - 1)) == 0 && (num - 1) % 3 == 0
}

File: python/342.py
# Should look like 10000, 1000000 in binary expressions.
class Solution(object):
    def isPowerOfFour(self, num):
        """
        :type num: int
        :rtype: bool
        """
        numBin = "{0:b}".format(num)
        if num < 0:
            return False
        if numBin == "1" + "0"*(len(numBin) - 1) and len(numBin) % 2 == 1:
            return True
        return False

# Found another solution https://www.hrwhisper.me/leetcode-power-four/.
return num > 0 and num & (num - 1) == 0 and num & 0x55555555 != 0
# 1. check num > 0
# 2. check 10000000...
# 3. check 1 is before even number of 0s, 0x5555... = 01010101...